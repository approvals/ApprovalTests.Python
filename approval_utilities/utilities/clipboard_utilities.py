def copy_to_clipboard(text: str) -> None:
    """
    This acts as a wrapper to defer the importing of pyperclip util actual usage
    """
    import pyperclip

    pyperclip.copy(text)
