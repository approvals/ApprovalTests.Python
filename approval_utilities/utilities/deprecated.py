import functools
import warnings
from typing import Callable


def deprecated(reason: str) -> Callable:
    def decorator(func: Callable) -> Callable:
        """This is a decorator which can be used to mark functions
        as deprecated. It will result in a warning being emitted
        when the function is used."""

        @functools.wraps(func)
        def new_func(*args, **kwargs):
            warnings.simplefilter("always", DeprecationWarning)  # turn off filter
            warnings.warn(
                f"Call to deprecated function {func.__name__}. {reason}",
                category=DeprecationWarning,
                stacklevel=2,
            )
            warnings.simplefilter("default", DeprecationWarning)  # reset filter
            return func(*args, **kwargs)

        return new_func

    return decorator
