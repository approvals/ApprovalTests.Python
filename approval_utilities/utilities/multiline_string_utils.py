import textwrap
from typing import Optional


def remove_indentation_from(text: Optional[str]) -> str:
    if not text:
        return ""
    cleaned = textwrap.dedent(text + "|")
    if cleaned.startswith("\n"):
        cleaned = cleaned[1:]
    return cleaned[:-1]
