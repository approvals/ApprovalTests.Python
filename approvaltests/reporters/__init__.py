"""Approvaltests.reporters module."""
from .clipboard_reporter import *
from .default_reporter_factory import *
from .diff_reporter import *
from .first_working_reporter import *
from .generic_diff_reporter import *
from .generic_diff_reporter_config import *
from .generic_diff_reporter_factory import *
from .multi_reporter import *
from .python_native_reporter import *
from .received_file_launcher_reporter import *
from .reporter_that_automatically_approves import *
from .report_all_to_clipboard import *
from .report_by_creating_diff_file import *
