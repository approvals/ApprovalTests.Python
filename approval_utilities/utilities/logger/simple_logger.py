import threading
from typing import Any, Callable, ContextManager, Optional, Union

from approval_utilities.utilities.logger.logging_instance import LoggingInstance
from approval_utilities.utilities.string_wrapper import StringWrapper
from approval_utilities.utilities.wrapper import SingleWrapper, ThreadedWrapper, Wrapper


class SimpleLogger:
    _wrapper: Wrapper[LoggingInstance] = SingleWrapper(LoggingInstance())

    @staticmethod
    def register_logger(log_method: Callable[[str], None]) -> None:
        SimpleLogger._wrapper.get().logger = log_method

    @staticmethod
    def log_to_string(log_separate_threads: bool = True) -> StringWrapper:
        with threading.Lock():
            if log_separate_threads and not isinstance(
                SimpleLogger._wrapper, ThreadedWrapper
            ):
                SimpleLogger._wrapper = ThreadedWrapper(lambda: LoggingInstance())
        return SimpleLogger._wrapper.get().log_to_string()

    @staticmethod
    def use_markers(
        parameter_text: Optional[Union[str, Callable[[], str]]] = None,
    ) -> ContextManager[None]:
        return SimpleLogger._wrapper.get().use_markers(
            parameter_text, additional_stack=1
        )

    @staticmethod
    def variable(name: str, value: Any, show_types: bool = False) -> None:
        SimpleLogger._wrapper.get().variable(name, value, show_types=show_types)

    @staticmethod
    def hour_glass() -> None:
        SimpleLogger._wrapper.get().hour_glass()

    @staticmethod
    def event(event_name: str) -> None:
        SimpleLogger._wrapper.get().event(event_name)

    @staticmethod
    def show_timestamps(display: bool) -> None:
        SimpleLogger._wrapper.get().log_with_timestamps = display

    @staticmethod
    def query(query_text: str) -> None:
        SimpleLogger._wrapper.get().query(query_text)

    @staticmethod
    def message(message: str) -> None:
        SimpleLogger._wrapper.get().message(message)

    @staticmethod
    def warning(
        text: Union[str, BaseException] = "", exception: Optional[BaseException] = None
    ) -> None:
        SimpleLogger._wrapper.get().warning(text, exception)

    @staticmethod
    def show_queries(show: bool) -> None:
        SimpleLogger._wrapper.get().show_queries(show)

    @staticmethod
    def show_all(show: bool) -> None:
        SimpleLogger._wrapper.get().show_all(show)

    @staticmethod
    def show_messages(show: bool) -> None:
        SimpleLogger._wrapper.get().show_messages(show)

    @staticmethod
    def show_variables(show: bool) -> None:
        SimpleLogger._wrapper.get().show_variables(show)

    @staticmethod
    def show_hour_glass(show: bool) -> None:
        SimpleLogger._wrapper.get().show_hour_glass(show)

    @staticmethod
    def show_markers(show: bool) -> None:
        SimpleLogger._wrapper.get().show_markers(show)

    @staticmethod
    def show_events(show: bool) -> None:
        SimpleLogger._wrapper.get().show_events(show)
