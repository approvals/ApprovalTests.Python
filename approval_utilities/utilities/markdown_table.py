from typing import TYPE_CHECKING, Any, Callable, Iterable

from typing_extensions import override

from approval_utilities.approvaltests.core.verifiable import Verifiable
from approval_utilities.approvaltests.core.verify_parameters import VerifyParameters

if TYPE_CHECKING:
    from approvaltests.core.options import Options


class MarkdownTable(Verifiable):
    def __init__(self) -> None:
        self.markdown: str = ""

    @override
    def get_verify_parameters(self, options: "Options") -> VerifyParameters:
        return VerifyParameters(options.for_file.with_extension(".md"))

    @staticmethod
    def with_headers(*column_names: str) -> "MarkdownTable":
        table = MarkdownTable()
        table.add_rows(*column_names)
        dividers = map(lambda _: "---", column_names)
        table.add_rows(*dividers)
        return table

    def add_rows(self, *column_names: str) -> "MarkdownTable":
        self.markdown += MarkdownTable.print_row(*column_names)
        return self

    @override
    def __str__(self) -> str:
        return self.markdown

    @staticmethod
    def print_row(*column_names: Any) -> str:
        row = "| " + " | ".join(map(str, column_names)) + " |"
        return row + "\n"

    def add_rows_for_inputs(
        self, inputs: Iterable[Any], *input_transformers: Callable[[Any], Any]
    ) -> "MarkdownTable":
        def transform_resolver_for_input(
            input_value: Any,
        ) -> Callable[[Callable[[Any], Any]], Any]:
            return lambda transform: transform(input_value)

        for row_input in inputs:
            row = [row_input]
            row += map(transform_resolver_for_input(row_input), input_transformers)
            self.add_rows(*row)
        return self
