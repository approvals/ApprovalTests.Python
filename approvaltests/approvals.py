from threading import local
import xml.dom.minidom

from approvaltests import to_json
from approvaltests.approval_exception import ApprovalException
from approvaltests.core.namer import Namer
from approvaltests.core.scenario_namer import ScenarioNamer
from approvaltests.core.pytest_namer import PytestNamer
from approvaltests.file_approver import FileApprover
from approvaltests.list_utils import format_list
from approvaltests.reporters.diff_reporter import DiffReporter
from approvaltests.string_writer import StringWriter

DEFAULT_REPORTER = local()


def set_default_reporter(reporter):
    global DEFAULT_REPORTER
    DEFAULT_REPORTER.v = reporter
    

def get_default_reporter():
    if not hasattr(DEFAULT_REPORTER, "v") or DEFAULT_REPORTER.v is None:
        return DiffReporter()
    return DEFAULT_REPORTER.v


def get_reporter(reporter):
    return reporter or get_default_reporter()


def get_default_namer():
    return Namer()


def verify(data, reporter=None, namer=None):
    reporter_to_use = reporter or get_default_reporter()
    namer_to_use = namer or get_default_namer()
    verify_with_namer(data, namer_to_use, reporter_to_use)


def verify_with_namer(data, namer, reporter):
    reporter = get_reporter(reporter)
    approver = FileApprover()
    writer = StringWriter(data)
    error = approver.verify(namer, writer, reporter)
    if error is not None:
        raise ApprovalException(error)


def verify_as_json(object, reporter=None):
    n_ = to_json(object) + "\n"
    verify(n_, reporter)


def verify_xml(xml_string, reporter=None):
    try:
        dom = xml.dom.minidom.parseString(xml_string)
        pretty_xml = dom.toprettyxml()
    except Exception:
        pretty_xml = xml_string
    verify_with_namer(pretty_xml, Namer(extension='.xml'), reporter)


def verify_file(file_name, reporter=None):
    with open(file_name, 'r') as f:
        file_contents = f.read()
    verify(file_contents, reporter)


def verify_all(header, alist, formatter=None, reporter=None):
    text = format_list(alist, formatter, header)
    verify(text, reporter)


def get_scenario_namer(scenario_name):
    return ScenarioNamer(get_default_namer(), scenario_name)

def get_pytest_namer(pytest_request):
    return PytestNamer(get_default_namer(), pytest_request)