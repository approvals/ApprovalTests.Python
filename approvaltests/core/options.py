from typing import Callable, Dict, Optional

from approvaltests.core.comparator import Comparator
from approvaltests.core.namer import Namer
from approvaltests.core.reporter import Reporter
from approvaltests.file_approver import FileComparator
from approvaltests.inline.inline_options import InlineOptions
from approvaltests.scrubbers import combine_scrubbers


class FileOptions:
    def __init__(self, fields: Dict):
        self.fields = fields

    @property
    def file_extention(self) -> str:
        return self.fields.get("extension_with_dot", ".txt")

    def with_extension(
        self,
        extension_with_dot: str,
        *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/,
        no_override: bool = False,
    ) -> "Options":
        if not extension_with_dot.startswith("."):
            extension_with_dot = "." + extension_with_dot
        if no_override and "extension_with_dot" in self.fields:
            extension_with_dot = self.fields["extension_with_dot"]
        return Options({**self.fields, "extension_with_dot": extension_with_dot})


class Options:
    def __init__(self, fields: Optional[Dict] = None):
        self.fields = fields or {}

    @property
    def reporter(self) -> Reporter:
        from approvaltests.reporters.default_reporter_factory import (
            get_default_reporter,
        )

        return self.fields.get("reporter", get_default_reporter())

    @property
    def comparator(self) -> Comparator:
        return self.fields.get("comparator", FileComparator())

    def with_comparator(self, comparator: Comparator) -> "Options":
        return Options({**self.fields, "comparator": comparator})

    def scrub(self, data: str) -> str:
        if self.has_scrubber():
            return self.fields["scrubber_func"](data)
        return data

    def with_scrubber(self, scrubber_func: Callable[[str], str]) -> "Options":
        return Options({**self.fields, "scrubber_func": scrubber_func})

    def add_scrubber(self, scrubber: Callable[[str], str]) -> "Options":
        if self.has_scrubber():
            scrubber = combine_scrubbers(self.fields["scrubber_func"], scrubber)
        return self.with_scrubber(scrubber)

    def has_scrubber(self) -> bool:
        return "scrubber_func" in self.fields

    def with_reporter(self, reporter: "Reporter") -> "Options":
        return Options({**self.fields, "reporter": reporter})

    def with_namer(self, namer: "Namer") -> "Options":
        return Options({**self.fields, "namer": namer})

    @property
    def for_file(self) -> FileOptions:
        return FileOptions(self.fields)

    @property
    def namer(self) -> Namer:
        from approvaltests.namer.default_name import get_default_namer

        namer = self.fields.get("namer", get_default_namer())
        if hasattr(namer, "set_extension"):
            namer.set_extension(self.for_file.file_extention)
        return namer

    def inline(self, inline_options: Optional[InlineOptions] = None) -> "Options":
        from approvaltests.namer.inline_comparator import InlineComparator

        return InlineComparator().register(self, inline_options)

    def add_reporter(self, additional_reporter: Reporter) -> "Options":
        from approvaltests.reporters import MultiReporter

        return self.with_reporter(MultiReporter(self.reporter, additional_reporter))
