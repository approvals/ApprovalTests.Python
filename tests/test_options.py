import importlib
import inspect
import sys

from approvaltests import (
    ReporterForTesting,
    approvals,
    combination_approvals,
    get_default_reporter,
    verify,
    verify_all,
)
from approvaltests.core.options import Options
from approvaltests.reporters import MultiReporter, ReportByCreatingDiffFile
from approvaltests.reporters.report_with_beyond_compare import ReportWithPycharm
from approvaltests.utilities import command_line_approvals
from approvaltests.utilities.logger import simple_logger_approvals
from approvaltests.utilities.logging import logging_approvals

_approvals_modules = list(
    sorted(
        filter(
            lambda name: name.startswith("approvaltests.")
            and name.endswith("approvals"),
            sys.modules.keys(),
        )
    )
)


def test_list_of_modules() -> None:
    verify_all("", _approvals_modules)


def test_every_function_in_approvals_with_verify_has_an_options() -> None:
    for module_name in _approvals_modules:
        assert_verify_methods_have_options(importlib.import_module(module_name))


def assert_verify_methods_have_options(module) -> None:
    for function_name, obj in module.__dict__.items():
        if "verify" not in function_name:
            continue

        if not callable(obj):
            continue

        # if it has a decorator, we need to get the original function
        if hasattr(obj, "__wrapped__"):
            obj = obj.__wrapped__

        argspec = inspect.getfullargspec(obj)
        has_options = "options" in argspec.kwonlyargs
        assert (
            has_options
        ), f"Missing Keyword only parameter `options` in {function_name}:\n\t{argspec}"


def test_empty_options_has_default_reporter() -> None:
    ##approvals.set_default_reporter(None)
    options = Options()
    assert options.reporter == get_default_reporter()


def test_with_reporter() -> None:
    testr = ReporterForTesting()
    options = Options().with_reporter(testr)
    try:
        verify("Data2", options=options)
    except:
        pass

    assert testr.called


def test_setting_reporter() -> None:
    testr = ReporterForTesting()
    options = Options().with_reporter(testr)
    assert options.reporter == testr


def test_file_extensions() -> None:
    approvals.settings().allow_multiple_verify_calls_for_this_method()
    content = "# This is a markdown header\n"
    # begin-snippet: options_with_file_extension
    verify(content, options=Options().for_file.with_extension(".md"))
    # end-snippet
    verify(content, options=Options().for_file.with_extension("md"))


def test_add_reporter() -> None:
    # current behaviour, override
    options0 = (
        Options()
        .with_reporter(ReportByCreatingDiffFile())
        .with_reporter(ReportWithPycharm())
    )
    assert type(options0.reporter) == ReportWithPycharm

    # current work around, create a MultiReporter
    options_multi = Options().with_reporter(
        MultiReporter(ReportByCreatingDiffFile(), ReportWithPycharm())
    )
    assert (
        str(options_multi.reporter)
        == "MultiReporter(ReportByCreatingDiffFile, ReportWithPycharm)"
    )

    # new behaviour, append
    options0 = (
        Options()
        .with_reporter(ReportByCreatingDiffFile())
        .add_reporter(ReportWithPycharm())
    )
    assert (
        str(options_multi.reporter)
        == "MultiReporter(ReportByCreatingDiffFile, ReportWithPycharm)"
    )
