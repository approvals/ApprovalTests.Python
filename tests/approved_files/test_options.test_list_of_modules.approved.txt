0) approvaltests.approvals
1) approvaltests.combination_approvals
2) approvaltests.utilities.command_line_approvals
3) approvaltests.utilities.logger.simple_logger_approvals
4) approvaltests.utilities.logging.logging_approvals
